File: flutter_app/lib/chat_screen.dart
import 'package:flutter/material.dart';
import 'dart:math';

class Message {
  final String id;
  final String text;
  final bool isFromUser;
  final String timestamp;

  Message({
    String? id,
    required this.text,
    required this.isFromUser,
    required this.timestamp,
  }) : id = id ?? UniqueKey().toString();
}

class ChatItem {
  final String id;
  final String name;
  final bool isAI;
  final List<Message> messages;

  ChatItem({
    String? id,
    required this.name,
    this.isAI = false,
    this.messages = const [],
  }) : id = id ?? UniqueKey().toString();
}

class ChatListScreen extends StatefulWidget {
  final List<ChatItem> chats;
  final void Function(ChatItem) onChatClick;
  final void Function(String, bool) onNewChat;
  const ChatListScreen({
    super.key,
    required this.chats,
    required this.onChatClick,
    required this.onNewChat,
  });

  @override
  State<ChatListScreen> createState() => _ChatListScreenState();
}

class _ChatListScreenState extends State<ChatListScreen> {
  bool showDialog = false;
  String search = '';

  @override
  Widget build(BuildContext context) {
    final filtered = widget.chats
        .where((c) => c.name.toLowerCase().contains(search.toLowerCase()))
        .toList();
    return Scaffold(
      appBar: AppBar(
        title: const Text('Messages', style: TextStyle(fontWeight: FontWeight.bold)),
        actions: [
          IconButton(
            icon: const Icon(Icons.add),
            onPressed: () => setState(() => showDialog = true),
          ),
        ],
      ),
      body: Column(
        children: [
          Padding(
            padding: const EdgeInsets.all(12.0),
            child: TextField(
              decoration: const InputDecoration(labelText: 'Search'),
              onChanged: (v) => setState(() => search = v),
            ),
          ),
          Expanded(
            child: ListView(
              children: filtered
                  .map((chat) => ChatRow(
                        chat: chat,
                        onClick: () => widget.onChatClick(chat),
                      ))
                  .toList(),
            ),
          ),
        ],
      ),
      floatingActionButton: showDialog
          ? null
          : FloatingActionButton(
              onPressed: () => setState(() => showDialog = true),
              child: const Icon(Icons.add),
            ),
    );
  }
}

class ChatRow extends StatelessWidget {
  final ChatItem chat;
  final VoidCallback onClick;
  const ChatRow({super.key, required this.chat, required this.onClick});

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: onClick,
      child: Padding(
        padding: const EdgeInsets.all(12.0),
        child: Row(
          children: [
            CircleAvatar(
              radius: 24,
              child: Text(
                chat.name.isNotEmpty ? chat.name[0].toUpperCase() : '?',
                style: const TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
              ),
            ),
            const SizedBox(width: 12),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(chat.name, style: const TextStyle(fontWeight: FontWeight.bold)),
                  Text(
                    chat.messages.isNotEmpty ? chat.messages.last.text : 'No messages yet',
                    maxLines: 1,
                    style: const TextStyle(color: Colors.grey),
                  ),
                ],
              ),
            ),
            Text(
              chat.messages.isNotEmpty ? chat.messages.last.timestamp : '',
              style: Theme.of(context).textTheme.bodySmall,
            ),
          ],
        ),
      ),
    );
  }
}

class ConversationScreen extends StatefulWidget {
  final ChatItem chat;
  final VoidCallback onBack;
  final void Function(String) onSendMessage;
  const ConversationScreen({
    super.key,
    required this.chat,
    required this.onBack,
    required this.onSendMessage,
  });

  @override
  State<ConversationScreen> createState() => _ConversationScreenState();
}

class _ConversationScreenState extends State<ConversationScreen> {
  final TextEditingController _controller = TextEditingController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.chat.name),
        leading: IconButton(
          icon: const Icon(Icons.arrow_back),
          onPressed: widget.onBack,
        ),
      ),
      body: Column(
        children: [
          Expanded(
            child: ListView(
              reverse: true,
              children: widget.chat.messages.reversed
                  .map((msg) => MessageBubble(message: msg))
                  .toList(),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Row(
              children: [
                Expanded(
                  child: TextField(
                    controller: _controller,
                    decoration: const InputDecoration(hintText: 'Type a message...'),
                  ),
                ),
                const SizedBox(width: 8),
                ElevatedButton(
                  onPressed: () {
                    if (_controller.text.trim().isNotEmpty) {
                      widget.onSendMessage(_controller.text.trim());
                      _controller.clear();
                    }
                  },
                  child: const Text('Send'),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class MessageBubble extends StatelessWidget {
  final Message message;
  const MessageBubble({super.key, required this.message});

  @override
  Widget build(BuildContext context) {
    final isUser = message.isFromUser;
    return Row(
      mainAxisAlignment: isUser ? MainAxisAlignment.end : MainAxisAlignment.start,
      children: [
        Card(
          color: isUser ? Theme.of(context).colorScheme.primaryContainer : Theme.of(context).colorScheme.secondaryContainer,
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
          child: Padding(
            padding: const EdgeInsets.all(12.0),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.end,
              children: [
                Text(message.text),
                Text(
                  message.timestamp,
                  style: Theme.of(context).textTheme.bodySmall?.copyWith(color: Colors.grey),
                ),
              ],
            ),
          ),
        ),
      ],
    );
  }
}

class NewConversationDialog extends StatefulWidget {
  final void Function(String, bool) onCreate;
  final VoidCallback onDismiss;
  const NewConversationDialog({super.key, required this.onCreate, required this.onDismiss});

  @override
  State<NewConversationDialog> createState() => _NewConversationDialogState();
}

class _NewConversationDialogState extends State<NewConversationDialog> {
  String name = '';
  bool isGroup = false;

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      title: const Text('New Conversation'),
      content: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          TextField(
            decoration: const InputDecoration(labelText: 'Name'),
            onChanged: (v) => setState(() => name = v),
          ),
          const SizedBox(height: 12),
          Row(
            children: [
              Expanded(
                child: ElevatedButton(
                  onPressed: () => setState(() => isGroup = false),
                  style: ElevatedButton.styleFrom(
                    backgroundColor: !isGroup ? Theme.of(context).colorScheme.primary : Theme.of(context).colorScheme.surfaceVariant,
                  ),
                  child: const Text('Individual'),
                ),
              ),
              const SizedBox(width: 8),
              Expanded(
                child: ElevatedButton(
                  onPressed: () => setState(() => isGroup = true),
                  style: ElevatedButton.styleFrom(
                    backgroundColor: isGroup ? Theme.of(context).colorScheme.primary : Theme.of(context).colorScheme.surfaceVariant,
                  ),
                  child: const Text('Group'),
                ),
              ),
            ],
          ),
        ],
      ),
      actions: [
        TextButton(onPressed: widget.onDismiss, child: const Text('Cancel')),
        ElevatedButton(
          onPressed: name.isNotEmpty ? () => widget.onCreate(name, isGroup) : null,
          child: const Text('Create'),
        ),
      ],
    );
  }
}

File: app/src/main/java/com/example/mind/ChatScreen.kt
package com.example.mind

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Add
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import java.util.UUID

// ---------- MODELS ----------

data class Message(
    val id: String = UUID.randomUUID().toString(),
    val text: String,
    val isFromUser: Boolean,
    val timestamp: String
)

data class ChatItem(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val isAI: Boolean = false,
    val messages: List<Message> = emptyList()
)

// ---------- CHAT LIST ----------

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ChatListScreen(
    chats: List<ChatItem>,
    onChatClick: (ChatItem) -> Unit,
    onNewChat: (String, Boolean) -> Unit
) {
    var showDialog by remember { mutableStateOf(false) }
    var search by remember { mutableStateOf("") }

    val filtered = chats.filter {
        it.name.contains(search, ignoreCase = true)
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Messages", fontWeight = FontWeight.Bold) },
                actions = {
                    IconButton(onClick = { showDialog = true }) {
                        Icon(Icons.Default.Add, contentDescription = "New Chat")
                    }
                }
            )
        }
    ) { padding ->
        Column(Modifier.padding(padding)) {
            OutlinedTextField(
                value = search,
                onValueChange = { search = it },
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(12.dp),
                label = { Text("Search") }
            )

            LazyColumn {
                items(filtered) { chat ->
                    ChatRow(chat) { onChatClick(chat) }
                }
            }
        }
    }

    if (showDialog) {
        NewConversationDialog(
            onDismiss = { showDialog = false },
            onCreate = { name, isGroup ->
                onNewChat(name, isGroup)
                showDialog = false
            }
        )
    }
}

@Composable
fun ChatRow(chat: ChatItem, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Box(
            modifier = Modifier
                .size(48.dp)
                .background(MaterialTheme.colorScheme.secondaryContainer, CircleShape),
            contentAlignment = Alignment.Center
        ) {
            Text(
                text = chat.name.first().uppercase(),
                fontSize = 20.sp,
                fontWeight = FontWeight.Bold
            )
        }

        Spacer(Modifier.width(12.dp))

        Column(Modifier.weight(1f)) {
            Text(chat.name, fontWeight = FontWeight.Bold)
            Text(
                chat.messages.lastOrNull()?.text ?: "No messages yet",
                maxLines = 1,
                color = Color.Gray
            )
        }

        Text(
            chat.messages.lastOrNull()?.timestamp ?: "",
            style = MaterialTheme.typography.bodySmall
        )
    }
}

// ---------- CONVERSATION ----------

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ConversationScreen(
    chat: ChatItem,
    onBack: () -> Unit,
    onSendMessage: (String) -> Unit
) {
    var input by remember { mutableStateOf("") }
    val listState = rememberLazyListState()

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text(chat.name) },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, "Back")
                    }
                }
            )
        },
        bottomBar = {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(8.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                OutlinedTextField(
                    value = input,
                    onValueChange = { input = it },
                    modifier = Modifier.weight(1f),
                    placeholder = { Text("Type a message...") }
                )
                Spacer(Modifier.width(8.dp))
                Button(
                    onClick = {
                        if (input.isNotBlank()) {
                            onSendMessage(input)
                            input = ""
                        }
                    }
                ) { Text("Send") }
            }
        }
    ) { padding ->
        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding),
            reverseLayout = true,
            state = listState
        ) {
            items(chat.messages.reversed()) {
                MessageBubble(it)
            }
        }
    }
}

@Composable
fun MessageBubble(message: Message) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(6.dp),
        horizontalArrangement =
            if (message.isFromUser) Arrangement.End else Arrangement.Start
    ) {
        Card(
            colors = CardDefaults.cardColors(
                containerColor =
                    if (message.isFromUser)
                        MaterialTheme.colorScheme.primaryContainer
                    else
                        MaterialTheme.colorScheme.secondaryContainer
            ),
            shape = RoundedCornerShape(16.dp)
        ) {
            Column(Modifier.padding(12.dp)) {
                Text(message.text)
                Text(
                    message.timestamp,
                    style = MaterialTheme.typography.bodySmall,
                    color = Color.Gray,
                    modifier = Modifier.align(Alignment.End)
                )
            }
        }
    }
}

// ---------- NEW CHAT DIALOG ----------

@Composable
fun NewConversationDialog(onDismiss: () -> Unit, onCreate: (String, Boolean) -> Unit
) {
    var name by remember { mutableStateOf("") }
    var isGroup by remember { mutableStateOf(false) }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("New Conversation") },
        text = {
            Column {
                OutlinedTextField(
                    value = name,
                    onValueChange = { name = it },
                    label = { Text("Name") }
                )
                Spacer(Modifier.height(12.dp))
                Row {
                    Button(
                        onClick = { isGroup = false },
                        modifier = Modifier.weight(1f),
                        colors = ButtonDefaults.buttonColors(
                            containerColor =
                                if (!isGroup)
                                    MaterialTheme.colorScheme.primary
                                else
                                    MaterialTheme.colorScheme.surfaceVariant
                        )
                    ) { Text("Individual") }

                    Spacer(Modifier.width(8.dp))

                    Button(
                        onClick = { isGroup = true },
                        modifier = Modifier.weight(1f),
                        colors = ButtonDefaults.buttonColors(
                            containerColor =
                                if (isGroup)
                                    MaterialTheme.colorScheme.primary
                                else
                                    MaterialTheme.colorScheme.surfaceVariant
                        )
                    ) { Text("Group") }
                }
            }
        },
        confirmButton = {
            Button(
                enabled = name.isNotBlank(),
                onClick = { onCreate(name, isGroup) }
            ) { Text("Create") }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) { Text("Cancel") }
        }
    )
}
